File: lab4/main.py
import cv2
import numpy as np

img = cv2.imread("media/img.png", cv2.IMREAD_GRAYSCALE)

kernel = np.ones((5,5),np.float32)/25
dst = cv2.filter2D(img,-1,kernel)

def apply_kernel(matr, ker):
    res = 0
    for i in range(len(matr)):
        for j in range(len(matr[i])):
            res += matr[i][j] * ker[i][j]
    return res

Gx = np.array([
    [-1,0,1],
    [-2,0,2],
    [-1,0,1]
])

Gy = np.array([
    [-1,-2,-1],
    [0,0,0],
    [1,2,1]
])

def quantize_angle(angle_deg):
    if 0 <= angle_deg < 22.5:
        return 2
    elif 22.5 <= angle_deg < 67.5:
        return 1
    elif 67.5 <= angle_deg < 112.5:
        return 0
    elif 112.5 <= angle_deg < 157.5:
        return 7
    elif 157.5 <= angle_deg < 202.5:
        return 6
    elif 202.5 <= angle_deg < 247.5:
        return 5
    elif 247.5 <= angle_deg < 292.5:
        return 4
    elif 292.5 <= angle_deg < 337.5:
        return 3
    else:
        return 2
    
new_img = np.zeros(img.shape)
grads = np.zeros(img.shape)

mxl = -1
for y in range(1, len(dst)-1):
    for x in range(1, len(dst[y])-1):
        gx = apply_kernel(img[y-1:y+2,x-1:x+2], Gx)
        gy = apply_kernel(img[y-1:y+2,x-1:x+2], Gy)

        l = np.sqrt(gx**2+gy**2)
        grads[y,x] = l
        mxl = max(mxl,l)

for y in range(1, len(dst)-1):
    for x in range(1, len(dst[y])-1):
        gx = apply_kernel(img[y-1:y+2,x-1:x+2], Gx)
        gy = apply_kernel(img[y-1:y+2,x-1:x+2], Gy)

        angle = np.rad2deg(np.arctan2(gy, gx))
        angle_round = quantize_angle(angle)

        if angle_round in [2,6] and grads[y,x]>max(grads[y,x-1],grads[y,x+1]):
            new_img[y,x] = 255
        elif angle_round in [0,4] and grads[y,x]>max(grads[y-1,x],grads[y+1,x]):
            new_img[y,x] = 255
        elif angle_round in [3,7] and grads[y,x]>max(grads[y-1,x-1],grads[y+1,x+1]):
            new_img[y,x] = 255
        elif angle_round in [1,5] and grads[y,x]>max(grads[y-1,x+1],grads[y+1,x-1]):
            new_img[y,x] = 255

low_level = mxl // 25
high_level = mxl // 10

borders = np.zeros(img.shape)

for y in range(1, len(new_img)-1):
    for x in range(1, len(new_img[y])-1):
        if new_img[y,x] >= high_level: borders[y,x] = 255
        elif low_level <= new_img[y,x] < high_level:
            if any(x >= high_level for x in [
                new_img[y-1,x],
                new_img[y+1,x],
                new_img[y,x-1],
                new_img[y,x+1],
                new_img[y-1,x-1],
                new_img[y-1,x+1],
                new_img[y+1,x-1],
                new_img[y+1,x+1]
            ]): borders[y,x] = 255

cv2.imshow('Display window1', dst)
cv2.imshow('Display window2', borders)

cv2.waitKey(0)
cv2.destroyAllWindows()

File: lab4/main.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <cmath>
using namespace cv;
using namespace std;

float apply_kernel(const Mat& matr, const Mat& ker) {
    float res = 0.0f;
    for (int i = 0; i < ker.rows; i++) {
        for (int j = 0; j < ker.cols; j++) {
            res += matr.at<uchar>(i, j) * ker.at<float>(i, j);
        }
    }
    return res;
}

int quantize_angle(float angle_deg) {
    if (0 <= angle_deg && angle_deg < 22.5) return 2;
    else if (22.5 <= angle_deg && angle_deg < 67.5) return 1;
    else if (67.5 <= angle_deg && angle_deg < 112.5) return 0;
    else if (112.5 <= angle_deg && angle_deg < 157.5) return 7;
    else if (157.5 <= angle_deg && angle_deg < 202.5) return 6;
    else if (202.5 <= angle_deg && angle_deg < 247.5) return 5;
    else if (247.5 <= angle_deg && angle_deg < 292.5) return 4;
    else if (292.5 <= angle_deg && angle_deg < 337.5) return 3;
    else return 2;
}

int main() {
    Mat img = imread("./img.png", IMREAD_GRAYSCALE);

    Mat kernel = Mat::ones(5, 5, CV_32F) / 25.0;
    Mat dst;
    filter2D(img, dst, -1, kernel);

    Mat Gx = (Mat_<float>(3,3) <<
        -1, 0, 1,
        -2, 0, 2,
        -1, 0, 1
    );
    Mat Gy = (Mat_<float>(3,3) <<
        -1, -2, -1,
         0,  0,  0,
         1,  2,  1
    );

    Mat new_img = Mat::zeros(img.size(), CV_32F);
    Mat grads = Mat::zeros(img.size(), CV_32F);

    float mxl = -1.0f;

    for (int y = 1; y < dst.rows - 1; y++) {
        for (int x = 1; x < dst.cols - 1; x++) {
            Mat roi = img(Rect(x-1, y-1, 3, 3));
            float gx = apply_kernel(roi, Gx);
            float gy = apply_kernel(roi, Gy);
            float l = sqrt(gx * gx + gy * gy);
            grads.at<float>(y, x) = l;
            if (l > mxl) mxl = l;
        }
    }

    for (int y = 1; y < dst.rows - 1; y++) {
        for (int x = 1; x < dst.cols - 1; x++) {
            Mat roi = img(Rect(x-1, y-1, 3, 3));
            float gx = apply_kernel(roi, Gx);
            float gy = apply_kernel(roi, Gy);

            float angle = atan2(gy, gx) * 180.0 / CV_PI;
            int angle_round = quantize_angle(angle);

            float val = grads.at<float>(y, x);
            if ((angle_round == 2 || angle_round == 6) && val > max(grads.at<float>(y, x-1), grads.at<float>(y, x+1)))
                new_img.at<float>(y, x) = 255;
            else if ((angle_round == 0 || angle_round == 4) && val > max(grads.at<float>(y-1, x), grads.at<float>(y+1, x)))
                new_img.at<float>(y, x) = 255;
            else if ((angle_round == 3 || angle_round == 7) && val > max(grads.at<float>(y-1, x-1), grads.at<float>(y+1, x+1)))
                new_img.at<float>(y, x) = 255;
            else if ((angle_round == 1 || angle_round == 5) && val > max(grads.at<float>(y-1, x+1), grads.at<float>(y+1, x-1)))
                new_img.at<float>(y, x) = 255;
        }
    }

    float low_level = mxl / 25.0f;
    float high_level = mxl / 10.0f;

    Mat borders = Mat::zeros(img.size(), CV_8U);

    for (int y = 1; y < new_img.rows - 1; y++) {
        for (int x = 1; x < new_img.cols - 1; x++) {
            float val = new_img.at<float>(y, x);
            if (val >= high_level) {
                borders.at<uchar>(y, x) = 255;
            } else if (val >= low_level && val < high_level) {
                bool strong_neighbor = false;
                for (int dy = -1; dy <= 1; dy++) {
                    for (int dx = -1; dx <= 1; dx++) {
                        if (new_img.at<float>(y+dy, x+dx) >= high_level)
                            strong_neighbor = true;
                    }
                }
                if (strong_neighbor) borders.at<uchar>(y, x) = 255;
            }
        }
    }

    imshow("Display window: Original", img);
    imshow("Display window: Gauss", dst);
    imshow("Display window: Borders", borders);

    waitKey(0);
    destroyAllWindows();
    return 0;
}
